mirgecom/inviscid.py

```diff
@@ -0,0 +1,105 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from mirgecom.fluid import (
+    split_conserved,
+    join_conserved
+)
+
+
+def inviscid_flux(discr, eos, q):
+    r"""Compute the inviscid flux vectors from flow solution *q*.
+
+    The inviscid fluxes are
+    $(\rho\vec{V},(\rho{E}+p)\vec{V},\rho(\vec{V}\otimes\vec{V})
+    +p\mathbf{I}, \rho{Y_s}\vec{V})$
+
+    .. note::
+
+        The fluxes are returned as a 2D object array with shape:
+        ``(num_equations, ndim)``.  Each entry in the
+        flux array is a :class:`~meshmode.dof_array.DOFArray`.  This
+        form and shape for the flux data is required by the built-in
+        state data handling mechanism in :mod:`mirgecom.fluid`. That
+        mechanism is used by at least
+        :class:`mirgecom.fluid.ConservedVars`, and
+        :func:`mirgecom.fluid.join_conserved`, and
+        :func:`mirgecom.fluid.split_conserved`.
+    """
+    dim = discr.dim
+    cv = split_conserved(dim, q)
+    p = eos.pressure(cv)
+
+    mom = cv.momentum
+
+    return join_conserved(dim,
+            mass=mom,
+            energy=mom * (cv.energy + p) / cv.mass,
+            momentum=np.outer(mom, mom) / cv.mass + np.eye(dim)*p,
+            species_mass=(  # reshaped: (nspecies, dim)
+                (mom / cv.mass) * cv.species_mass.reshape(-1, 1)))
+
+
+def get_inviscid_timestep(discr, eos, cfl, q):
+    """Routine (will) return the (local) maximum stable inviscid timestep.
+
+    Currently, it's a hack waiting for the geometric_factor helpers port
+    from grudge.
+    """
+    dim = discr.dim
+    mesh = discr.mesh
+    order = max([grp.order for grp in discr.discr_from_dd("vol").groups])
+    nelements = mesh.nelements
+    nel_1d = nelements ** (1.0 / (1.0 * dim))
+
+    # This roughly reproduces the timestep AK used in wave toy
+    dt = (1.0 - 0.25 * (dim - 1)) / (nel_1d * order ** 2)
+    return cfl * dt
+
+#    dt_ngf = dt_non_geometric_factor(discr.mesh)
+#    dt_gf  = dt_geometric_factor(discr.mesh)
+#    wavespeeds = compute_wavespeed(w,eos=eos)
+#    max_v = clmath.max(wavespeeds)
+#    return c*dt_ngf*dt_gf/max_v
```

Comment:
I don't know what's going on here, but I like the commented-out code way more :smile: Are there any issues related to extracting geometric factors and the like? I honestly think this should probably become a higher priority since adaptive time-stepping and CFL monitoring (which @anderson2981 mentioned before) will need these.

@inducer any sense for what is missing / what work is involved to get this information?